source/common/network/win32_socket_handle_impl.cc

```diff
@@ -71,10 +79,39 @@ Api::IoCallUint64Result Win32SocketHandleImpl::recvmmsg(RawSliceArrays& slices,
 }
 
 Api::IoCallUint64Result Win32SocketHandleImpl::recv(void* buffer, size_t length, int flags) {
-
-  Api::IoCallUint64Result result = IoSocketHandleImpl::recv(buffer, length, flags);
-  reEnableEventBasedOnIOResult(result, Event::FileReadyType::Read);
-  return result;
+  if (flags & MSG_PEEK) {
+    // can a remote OOM us now that we are not protected by readDisable?
+    Api::IoCallUint64Result peek_result = drainToPeekBuffer();
+    //  Some fatal error happened
+    if (!peek_result.wouldBlock()) {
+      return peek_result;
+    }
+
+    // No data available, register read again.
+    if (peek_result.wouldBlock() && peek_buffer_->length() == 0) {
+      if (file_event_) {
+        file_event_->registerEventIfEmulatedEdge(Event::FileReadyType::Read);
+      }
+      return peek_result;
+    }
+
+    Api::IoCallUint64Result result = peekFromPeekBuffer(buffer, length);
+    if (peek_buffer_->length() < length && file_event_) {
+      file_event_->registerEventIfEmulatedEdge(Event::FileReadyType::Read);
+    } else {
+      // This means that our peak buffer has more data than what the user
+      // wanted. Return the slice to the caller.
+      // How can the caller (v2 proxy protocol inspector) reactivate the events again here?
+    }
+    return result;
+  }
+  if (peek_buffer_->length() == 0) {
+    Api::IoCallUint64Result result = IoSocketHandleImpl::recv(buffer, length, flags);
+    reEnableEventBasedOnIOResult(result, Event::FileReadyType::Read);
+    return result;
+  } else {
+    return readFromPeekBuffer(buffer, length);
```

Comment:
I guess we should re-enable the event here also, after read from peek buffer, we want to continue the read.